test/realm_test.dart

```diff
@@ -242,6 +242,26 @@ Future<void> main([List<String>? args]) async {
     expect(() => realm.write(() => realm.add(carTwo)), throws<RealmException>());
   });
 
+  test('Realm adding objects with duplicate primary with update flag', () {
+    var config = Configuration.local([Car.schema]);
+    var realm = getRealm(config);
+
+    final carOne = Car("Toyota");
+    final carTwo = Car("Toyota");
+    realm.write(() => realm.add(carOne));
+    expect(realm.write(() => realm.add(carTwo, update: true)), carTwo);
```

Comment:
Should this compare with `carOne`?

```suggestion
    expect(realm.write(() => realm.add(carTwo, update: true)), carOne);
```

Reply:
No

Reply:
So what is this asserting then? That adding an object to the Realm returns the same object?

Reply:
This test would throw an exception until the update flag was correctly working, independent of testing against car1 or car2. I can see why this test reads less naturally now, but it was added as part of a TDD flow. I can revisit the tests, but I risk changing them in the name of readability in a way that may not capture the original error anymore.

Anyway, I will do that, as I can see you have other comments to the same effect.

Reply:
If we want to assert that something doesn't throw, then a good practice would be to use the `returnsNormally` matcher as that better conveys the effect being tested. There's nothing wrong with adding preliminary tests as part of a TDD flow, but my suggestion would be to revisit and update them once the feature gets ready for shipping. Otherwise you're risking tying your tests to the actual implementation (like the example I gave above - a refactoring of the method to do nothing if `update: true` will not be caught by this test).

Reply:
I will.. tomorrow 😪